rosidl_typesupport_opensplice_c/resource/msg__type_support_c.cpp.template

```diff
@@ -0,0 +1,552 @@
+  }
+  if (!ros_message) {
+    return "ros message handle is null";
+  }
+
+  DDS::DataWriter * topic_writer = static_cast<DDS::DataWriter *>(dds_data_writer);
+
+  __dds_msg_type dds_message;
+  const char * err_msg = convert_ros_to_dds(ros_message, &dds_message);
+  if (err_msg != 0) {
+    return err_msg;
+  }
+
+  @(__dds_msg_type_prefix)DataWriter * data_writer =
+    @(__dds_msg_type_prefix)DataWriter::_narrow(topic_writer);
+  DDS::ReturnCode_t status = data_writer->write(dds_message, DDS::HANDLE_NIL);
+@[for field in spec.fields]@
+@[if field.type.type == 'string']@
+@[if field.type.is_array]@
+  {
+@[if field.type.array_size]@
+    size_t size = @(field.type.array_size);
+@[else]@
+    size_t size = dds_message.@(field.name)_.length();
+@[end if]@
+    for (DDS::ULong i = 0; i < size; ++i) {
+      // This causes the DDS::String_mgr to release the given c string without freeing it.
+      dds_message.@(field.name)_[i]._retn();
+    }
+  }
+@[else]@
+  // This causes the DDS::String_mgr to release the given c string without freeing it.
+  dds_message.@(field.name)_._retn();
+@[end if]@
+@[end if]@
+@[end for]@
+  switch (status) {
+    case DDS::RETCODE_ERROR:
+      return "@(__dds_msg_type_prefix)DataWriter.write: "
+             "an internal error has occurred";
+    case DDS::RETCODE_BAD_PARAMETER:
+      return "@(__dds_msg_type_prefix)DataWriter.write: "
+             "bad handle or instance_data parameter";
+    case DDS::RETCODE_ALREADY_DELETED:
+      return "@(__dds_msg_type_prefix)DataWriter.write: "
+             "this @(__dds_msg_type_prefix)DataWriter has already been deleted";
+    case DDS::RETCODE_OUT_OF_RESOURCES:
+      return "@(__dds_msg_type_prefix)DataWriter.write: "
+             "out of resources";
+    case DDS::RETCODE_NOT_ENABLED:
+      return "@(__dds_msg_type_prefix)DataWriter.write: "
+             "this @(__dds_msg_type_prefix)DataWriter is not enabled";
+    case DDS::RETCODE_PRECONDITION_NOT_MET:
+      return "@(__dds_msg_type_prefix)DataWriter.write: "
+             "the handle has not been registered with this @(__dds_msg_type_prefix)DataWriter";
+    case DDS::RETCODE_TIMEOUT:
+      return "@(__dds_msg_type_prefix)DataWriter.write: "
+             "writing resulted in blocking and then exceeded the timeout set by the "
+             "max_blocking_time of the ReliabilityQosPolicy";
+    case DDS::RETCODE_OK:
+      return nullptr;
+    default:
+      return "@(__dds_msg_type_prefix)DataWriter.write: unknown return code";
+  }
+}
+
+static const char *
+convert_dds_to_ros(const void * untyped_dds_message, void * untyped_ros_message)
+{
+  if (!untyped_ros_message) {
+    return "ros message handle is null";
+  }
+  if (!untyped_dds_message) {
+    return "dds message handle is null";
+  }
+  const __dds_msg_type * dds_message = static_cast<const __dds_msg_type *>(untyped_dds_message);
+  __ros_msg_type * ros_message = static_cast<__ros_msg_type *>(untyped_ros_message);
+  (void)dds_message;
```

Comment:
I think I couldn't do that before because of some `goto` logic, but I'm no longer using `goto` in the convert functions. I'll try to move it back, but I'll have to start new CI.